File: 05-real-tools/src/mock.ts
/**
 * Mock 模式：演示真实工具调用流程
 *
 * 模拟 Agent 接收到用户请求后的完整工具调用链:
 *   1. Glob 查找文件
 *   2. Grep 搜索内容
 *   3. Read 读取文件
 *   4. Write 写入修改
 */

import type { Message, ContentBlock, Tool } from "./types.js";

let callCount = 0;

/** 模拟 5 轮工具调用链 */
export async function mockCallModel(
  _messages: Message[],
  _tools: Tool[],
): Promise<ContentBlock[]> {
  callCount++;

  // 如果是摘要请求
  if (_messages.length > 0) {
    const first = _messages[0];
    if (first.role === "system" && first.content[0]?.type === "text") {
      if (first.content[0].text.includes("摘要")) {
        return [{ type: "text", text: "对话摘要：Agent使用工具链分析了项目结构。" }];
      }
    }
  }

  // 模拟工具调用链：Glob → Grep → Read → Write → 总结
  if (callCount === 1) {
    return [
      { type: "text", text: "我先找一下项目里有哪些 TypeScript 文件。" },
      {
        type: "tool_use",
        id: "call_glob_001",
        name: "Glob",
        input: { pattern: "**/*.ts", path: "." },
      },
    ];
  }

  if (callCount === 2) {
    return [
      { type: "text", text: "找到了一些文件，我来搜索一下有没有 export function 的定义。" },
      {
        type: "tool_use",
        id: "call_grep_001",
        name: "Grep",
        input: { pattern: "export function", path: "." },
      },
    ];
  }

  if (callCount === 3) {
    return [
      { type: "text", text: "看到 types.ts 里有几个类型定义，我读取一下完整内容。" },
      {
        type: "tool_use",
        id: "call_read_001",
        name: "Read",
        input: { file_path: "src/types.ts" },
      },
    ];
  }

  if (callCount === 4) {
    return [
      { type: "text", text: "我注意到项目中缺少错误处理的示例，我来创建一个。" },
      {
        type: "tool_use",
        id: "call_write_001",
        name: "Write",
        input: {
          file_path: "/tmp/mock-claude-safe-loop.ts",
          content: "// 这是模拟写入的文件内容\nexport function safeLoop() {\n  try {\n    // loop logic\n  } catch (e) {\n    console.error(e);\n  }\n}\n",
        },
      },
    ];
  }

  // 第 5 轮：总结，结束循环
  return [
    {
      type: "text",
      text: `分析完毕。项目有多个 TypeScript 文件，我创建了一个带错误处理的示例文件。这次调用了 4 个工具：Glob（找文件）→ Grep（搜内容）→ Read（读文件）→ Write（写修改）。这就是 Claude Code 的真实工作流。`,
    },
  ];
}

/** 保存记忆的 mock（沿用） */
const saveMemoryTool: Tool = {
  name: "SaveMemory",
  description: "保存一条记忆",
  input_schema: {
    type: "object",
    properties: {
      name: { type: "string" },
      description: { type: "string" },
      type: { type: "string", enum: ["user", "feedback", "project", "reference"] },
      content: { type: "string" },
    },
    required: ["name", "description", "type", "content"],
  },
  riskLevel: "safe",
  execute: async (_input) => "记忆保存成功",
};

export const mockTools: Tool[] = [saveMemoryTool];

export function resetMockCounter() {
  callCount = 0;
}

File: 06-rust-rewrite/src/mock.rs
/// Mock 模型——模拟Agent的5轮决策
///
/// 对照 TS 版 mock.ts
/// 工具调用链: Glob → Grep → Read → Write → 总结

use async_trait::async_trait;
use serde_json::json;

use crate::model::ModelCaller;
use crate::types::{ChatMessage, ContentBlock};

/// Mock 调用器——实现 ModelCaller trait
///
/// 对照 TS 版：mock 模式直接按 turn 返回预设结果
pub struct MockCaller;

#[async_trait]
impl ModelCaller for MockCaller {
    async fn call(&self, messages: &[ChatMessage]) -> Vec<ContentBlock> {
        // 从消息数推算轮次：初始1条，每轮+2条
        let turn = (messages.len() + 1) / 2;
        mock_call_model(turn)
    }
}

pub fn mock_call_model(turn: usize) -> Vec<ContentBlock> {
    match turn {
        // 第1轮：Glob找文件
        1 => vec![
            ContentBlock::Text {
                text: "我先找一下项目里有哪些Rust文件。".into(),
            },
            ContentBlock::ToolUse {
                id: "call_glob_001".into(),
                name: "Glob".into(),
                input: json!({"pattern": "**/*.rs", "path": "."}),
            },
        ],
        // 第2轮：Grep搜函数定义
        2 => vec![
            ContentBlock::Text {
                text: "找到了一些文件，我来搜索函数定义。".into(),
            },
            ContentBlock::ToolUse {
                id: "call_grep_001".into(),
                name: "Grep".into(),
                input: json!({"pattern": "pub (fn|struct|enum)", "path": "."}),
            },
        ],
        // 第3轮：Read读文件
        3 => vec![
            ContentBlock::Text {
                text: "看到types.rs里有些类型定义，我读取一下。".into(),
            },
            ContentBlock::ToolUse {
                id: "call_read_001".into(),
                name: "Read".into(),
                input: json!({"file_path": "src/types.rs"}),
            },
        ],
        // 第4轮：Write写入
        4 => vec![
            ContentBlock::Text {
                text: "我给项目加一个输出示例文件。".into(),
            },
            ContentBlock::ToolUse {
                id: "call_write_001".into(),
                name: "Write".into(),
                input: json!({
                    "file_path": "/tmp/rust-agent-demo-output.txt",
                    "content": "# Rust Agent Demo Output\n\n这是Rust版Agent的输出示例。\n4个工具全部真实执行。\n\n工具调用链: Glob -> Grep -> Read -> Write\n"
                }),
            },
        ],
        // 第5轮：总结
        _ => vec![ContentBlock::Text {
            text: "分析完毕。调用了4个工具：Glob（找文件）→ Grep（搜内容）→ Read（读文件）→ Write（写输出）。这就是Rust版Agent的完整工作流。".into(),
        }],
    }
}
